src/training/communicator_mpi.h

```diff
@@ -0,0 +1,324 @@
+          } else {
+            mpi_->bCast(shard->data(), 
+                        shard->size(), 
+                        floatType(shard), 
+                        0);
+          }
+        }
+      }
+    }
+  }
+
+  // Distribute a single CPU-side io::Item to shards across multiple devices and MPI processes.
+  // This is used when restoring optimizer state, which is sharded.
+  // It is assumed that all MPI processes get the same data() passed. Hence, no MPI transfers are needed here.
+  void scatterState(const io::Item& data, const OptimizerBase::ScatterStateSetFunc& setFn) const override {
+    size_t dataSize = data.size();
+    size_t numShards = shardingMode_ == ShardingMode::global ? numRanks() : numLocalRanks(); // @TODO: numShards() function
+    size_t shardSize = (dataSize + numShards - 1) / numShards;
+    for(size_t localDeviceIndex = 0; localDeviceIndex < graphs_.size(); localDeviceIndex++) {
+      // We only slice out data that is kept in our MPI process. Remember that all MPI processes receive the same, complete data.
+      auto ncclRank = shardingMode_ == ShardingMode::global ? myRank(localDeviceIndex) : myLocalRank(localDeviceIndex);
+      size_t begin = ncclRank * shardSize;
+      size_t end   = std::min(begin + shardSize, dataSize);
+      setFn(localDeviceIndex, data.bytes.data() + begin, data.bytes.data() + end);
+    }
+  }
+
+  // Collect shards across multiple devices and MPI processes in the NCCL configuration into a single CPU-side io::Item.
+  // This is used when persisting optimizer state which is sharded.
+  io::Item gatherState(const OptimizerBase::GatherStateGetFunc& getFn) const override {
+    // first, concatenate over all local devices
+    io::Item localData = getFn(0);
+    for(size_t localDeviceIndex = 1; localDeviceIndex < graphs_.size(); localDeviceIndex++) {
+      localData.append(getFn(localDeviceIndex));
+    }
+    // localData now contains a concatentation of all local data
+
+    // second, concatenate across MPI processes
+    // Note that all local devices occupy consecutive ncclRanks in order.
+    io::Item data;
+    if (mpi_ && shardingMode_ == ShardingMode::global) {
+      io::Item tmp = localData; // temp buffer used multiple times; assign localData for initialization
+      // push one rank's data at a time using broadcast
+      for(size_t mpiRank = 0; mpiRank < mpi_->numMPIProcesses(); mpiRank++) {
+        // broadcast mpiRank's localData to all
+        if(mpiRank == mpi_->myMPIRank()) {
+          tmp = localData;
+        }
+        mpi_->bCast(tmp, /*rootRank=*/mpiRank);
+        // now all ranks have the same slice: concatenate (we will end up with the same on all MPI processes)
+        if(mpiRank == 0)
+          data = tmp;
+        else
+          data.append(tmp);
+      }
+    }
+    else { // no MPI: localData is the complete result already
+      data = std::move(localData);
+    }
+    return data;
+  }
+};
+
+}  // namespace marian
```

Comment:
Generally a ton of code duplication with NCCL communicator. Once you think about adding local sharding, might be worth to combine? Will make it easier to maintain and make sure that it does not go down the path of the other MPI implementations that ended up not being used and removed in the end.